hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -842,6 +847,36 @@ class Emit[C](
               deepCopy = false)
           }
 
+      case x@InsertFields(old, fields, fieldOrder) =>
+        if (fields.isEmpty)
+          emitI(old)
+        else
+          old.pType match {
+            case _: PStruct =>
+              val codeOld = emitI(old)
+              val updateMap = Map(fields: _*)
+
+              codeOld.map(cb) { oldPC =>
+                val oldPV = oldPC.asBaseStruct.memoize(cb, "insert_fields_old")
+
+                val itemsEC = x.pType.fields.map { f =>
+                  updateMap.get(f.name) match {
+                    case Some(vir) =>
+                      EmitCode.fromI(mb)(emitInNewBuilder(_, vir))
+                    case None =>
+                      EmitCode.fromI(mb)(oldPV.loadField(_, f.index).asInstanceOf[IEmitCode])
+                  }
+                }
+
+                x.pType.asInstanceOf[PCanonicalBaseStruct]
+                 .constructFromFields(cb, region.code, itemsEC, deepCopy = false)
+                 .asPCode
+              }
+            case _ =>
```

Comment:
this seems like dead code -- isn't the old ptype always a struct?

Reply:
I figured this was to handle InsertFields on a tuple. But you're right, we don't allow that in type checking/inference. I'll delete.